File: tests/Aiko.Core.Tests/EnvironmentSettingsTests.cs
namespace Aiko.Core.Tests;

public class EnvironmentSettingsTests
{
    private const string TwoEnvironments = """
        {
          "ringEnvironment": "Personal",
          "environments": [
            { "name": "Personal", "configDirectories": ["C:\\Users\\slayu\\.claude-personal"], "directMode": false },
            { "name": "Work", "configDirectories": ["C:\\Users\\slayu\\.claude"], "directMode": true }
          ]
        }
        """;

    [Fact]
    public void Reads_environments_with_their_folders_and_direct_mode()
    {
        var settings = EnvironmentSettings.FromJson(TwoEnvironments);

        Assert.Equal(2, settings.Environments.Count);
        Assert.Equal("Personal", settings.Environments[0].Name);
        Assert.False(settings.Environments[0].DirectMode);
        Assert.True(settings.Environments[1].DirectMode);
        Assert.Equal(@"C:\Users\slayu\.claude", settings.Environments[1].ConfigDirectories.Single());
    }

    [Fact]
    public void The_ring_shows_the_chosen_environment_and_the_dot_the_other_one()
    {
        var settings = EnvironmentSettings.FromJson(TwoEnvironments);

        Assert.Equal("Personal", settings.Ring!.Name);
        Assert.Equal("Work", settings.Dot!.Name);
    }

    [Fact]
    public void A_click_swaps_the_ring_and_the_dot()
    {
        var swapped = EnvironmentSettings.FromJson(TwoEnvironments).SwapRing();

        Assert.Equal("Work", swapped.Ring!.Name);
        Assert.Equal("Personal", swapped.Dot!.Name);
    }

    [Fact]
    public void With_one_environment_there_is_no_dot_and_nothing_to_swap()
    {
        var settings = EnvironmentSettings.FromJson("""
            { "environments": [ { "name": "Personal", "configDirectories": ["C:\\x"] } ] }
            """);

        Assert.Equal("Personal", settings.Ring!.Name);
        Assert.Null(settings.Dot);
        Assert.Equal("Personal", settings.SwapRing().Ring!.Name);
    }

    [Fact]
    public void An_unknown_ring_name_falls_back_to_the_first_environment()
    {
        var settings = EnvironmentSettings.FromJson("""
            {
              "ringEnvironment": "Renamed away",
              "environments": [ { "name": "Personal", "configDirectories": ["C:\\x"] } ]
            }
            """);

        Assert.Equal("Personal", settings.Ring!.Name);
    }

    [Fact]
    public void Environments_without_a_name_or_a_folder_are_dropped()
    {
        var settings = EnvironmentSettings.FromJson("""
            {
              "environments": [
                { "name": "", "configDirectories": ["C:\\x"] },
                { "name": "No folders", "configDirectories": [] },
                { "name": "Good", "configDirectories": ["C:\\x"] }
              ]
            }
            """);

        Assert.Single(settings.Environments);
        Assert.Equal("Good", settings.Environments[0].Name);
    }

    [Fact]
    public void What_is_written_can_be_read_back()
    {
        var settings = EnvironmentSettings.FromJson(TwoEnvironments).SwapRing();

        var again = EnvironmentSettings.FromJson(settings.ToJson());

        Assert.Equal("Work", again.Ring!.Name);
        Assert.Equal(2, again.Environments.Count);
        Assert.True(again.Environments.Single(e => e.Name == "Work").DirectMode);
    }

    [Theory]
    [InlineData("")]
    [InlineData("not json")]
    [InlineData("{}")]
    [InlineData("{ \"environments\": [] }")]
    public void Anything_unusable_means_no_environments_yet(string json)
    {
        var settings = EnvironmentSettings.FromJson(json);

        Assert.False(settings.HasEnvironments);
        Assert.Null(settings.Ring);
    }

    [Fact]
    public void A_file_from_before_the_persona_reads_with_the_persona_off()
    {
        var settings = EnvironmentSettings.FromJson(TwoEnvironments);

        Assert.All(settings.Environments, e => Assert.False(e.Persona));
    }

    [Fact]
    public void The_persona_flag_survives_a_trip_through_the_file()
    {
        var settings = EnvironmentSettings.FromJson(TwoEnvironments);
        var withPersona = settings with
        {
            Environments = [settings.Environments[0] with { Persona = true }, settings.Environments[1]],
        };

        var json = withPersona.ToJson();
        var back = EnvironmentSettings.FromJson(json);

        Assert.True(back.Environments[0].Persona);
        Assert.False(back.Environments[1].Persona);
        Assert.Contains($"\"schemaVersion\": {EnvironmentSettings.CurrentSchema}", json);
    }
}

File: mac/Tests/AikoKitTests/EnvironmentSettingsTests.swift
import Testing

@testable import AikoKit

struct EnvironmentSettingsTests {
    static let twoEnvironments = """
        {
          "ringEnvironment": "Personal",
          "environments": [
            { "name": "Personal", "configDirectories": ["C:\\\\Users\\\\slayu\\\\.claude-personal"], "directMode": false },
            { "name": "Work", "configDirectories": ["C:\\\\Users\\\\slayu\\\\.claude"], "directMode": true }
          ]
        }
        """

    @Test
    func readsEnvironmentsWithTheirFoldersAndDirectMode() {
        let settings = EnvironmentSettings.fromJson(Self.twoEnvironments)

        #expect(settings.environments.count == 2)
        #expect(settings.environments[0].name == "Personal")
        #expect(!settings.environments[0].directMode)
        #expect(settings.environments[1].directMode)
        #expect(settings.environments[1].configDirectories == [#"C:\Users\slayu\.claude"#])
    }

    @Test
    func theRingShowsTheChosenEnvironmentAndTheDotTheOtherOne() {
        let settings = EnvironmentSettings.fromJson(Self.twoEnvironments)

        #expect(settings.ring?.name == "Personal")
        #expect(settings.dot?.name == "Work")
    }

    @Test
    func aClickSwapsTheRingAndTheDot() {
        let swapped = EnvironmentSettings.fromJson(Self.twoEnvironments).swapRing()

        #expect(swapped.ring?.name == "Work")
        #expect(swapped.dot?.name == "Personal")
    }

    @Test
    func withOneEnvironmentThereIsNoDotAndNothingToSwap() {
        let settings = EnvironmentSettings.fromJson(
            #"{ "environments": [ { "name": "Personal", "configDirectories": ["C:\\x"] } ] }"#)

        #expect(settings.ring?.name == "Personal")
        #expect(settings.dot == nil)
        #expect(settings.swapRing().ring?.name == "Personal")
    }

    @Test
    func anUnknownRingNameFallsBackToTheFirstEnvironment() {
        let settings = EnvironmentSettings.fromJson(
            """
            {
              "ringEnvironment": "Renamed away",
              "environments": [ { "name": "Personal", "configDirectories": ["C:\\\\x"] } ]
            }
            """)

        #expect(settings.ring?.name == "Personal")
    }

    @Test
    func environmentsWithoutANameOrAFolderAreDropped() {
        let settings = EnvironmentSettings.fromJson(
            """
            {
              "environments": [
                { "name": "", "configDirectories": ["C:\\\\x"] },
                { "name": "No folders", "configDirectories": [] },
                { "name": "Good", "configDirectories": ["C:\\\\x"] }
              ]
            }
            """)

        #expect(settings.environments.count == 1)
        #expect(settings.environments[0].name == "Good")
    }

    @Test
    func whatIsWrittenCanBeReadBack() {
        let settings = EnvironmentSettings.fromJson(Self.twoEnvironments).swapRing()

        let again = EnvironmentSettings.fromJson(settings.toJson())

        #expect(again.ring?.name == "Work")
        #expect(again.environments.count == 2)
        #expect(again.environments.first { $0.name == "Work" }?.directMode == true)
    }

    @Test(arguments: ["", "not json", "{}", #"{ "environments": [] }"#])
    func anythingUnusableMeansNoEnvironmentsYet(json: String) {
        let settings = EnvironmentSettings.fromJson(json)

        #expect(!settings.hasEnvironments)
        #expect(settings.ring == nil)
    }

    @Test
    func aFileFromBeforeThePersonaReadsWithThePersonaOff() {
        let settings = EnvironmentSettings.fromJson(Self.twoEnvironments)

        #expect(settings.environments.allSatisfy { !$0.persona })
    }

    @Test
    func thePersonaFlagSurvivesATripThroughTheFile() {
        var settings = EnvironmentSettings.fromJson(Self.twoEnvironments)
        settings.environments[0].persona = true

        let json = settings.toJson()
        let back = EnvironmentSettings.fromJson(json)

        #expect(back.environments[0].persona)
        #expect(!back.environments[1].persona)
        #expect(json.contains("\"schemaVersion\": \(EnvironmentSettings.currentSchema)"))
    }
}
